src/filter.rs

```diff
@@ -59,61 +59,123 @@ pub fn unfilter(
-                    );
-                }
-                Ok(())
+            for i in 0..bpp {
+                current[i] = current[i].wrapping_add(previous[i] / 2);
+            }
+
+            macro_rules! avg_tail {
+                ($name:ident, $bpp:expr) => {
+                    fn $name(current: &mut [u8], previous: &[u8]) {
+                        let len = current.len();
+                        let current = &mut current[..len];
+                        let previous = &previous[..len];
+
+                        let mut current = current.chunks_exact_mut($bpp);
+                        let mut previous = previous.chunks_exact($bpp);
+
+                        let mut lprevious = current.next().unwrap();
+                        let _ = previous.next();
+
+                        while let Some(pprevious) = previous.next() {
+                            let pcurrent = current.next().unwrap();
+
+                            for i in 0..$bpp {
+                                let lprev = lprevious[i];
+                                let pprev = pprevious[i];
+                                pcurrent[i] = pcurrent[i].wrapping_add(
+                                    ((u16::from(lprev) + u16::from(pprev)) / 2) as u8,
+                                );
+                            }
+
+                            lprevious = pcurrent;
+                        }
+                    }
+                };
             }
+
+            avg_tail!(avg_tail_8, 8);
+            avg_tail!(avg_tail_6, 6);
+            avg_tail!(avg_tail_4, 4);
+            avg_tail!(avg_tail_3, 3);
+            avg_tail!(avg_tail_2, 2);
+            avg_tail!(avg_tail_1, 1);
+
+            match bpp {
+                8 => avg_tail_8(current, previous),
+                6 => avg_tail_6(current, previous),
+                4 => avg_tail_4(current, previous),
+                3 => avg_tail_3(current, previous),
+                2 => avg_tail_2(current, previous),
+                1 => avg_tail_1(current, previous),
+                _ => unreachable!("Invalid bytes per pixel"),
+            }
+
+            Ok(())
         }
         Paeth => {
-            if previous.len() < len {
-                Err("Filtering failed: not enough data in previous row")
-            } else if bpp > len {
-                Err("Filtering failed: bytes per pixel is greater than length of row")
-            } else {
-                for i in 0..bpp {
-                    current[i] = current[i].wrapping_add(filter_paeth(0, previous[i], 0));
-                }
+            let current = &mut current[..len];
+            let previous = require_length(previous, len)?;
+            if bpp > len {
+                return Err("Filtering failed: bytes per pixel is greater than length of row");
+            }
+
+            for i in 0..bpp {
+                current[i] = current[i].wrapping_add(filter_paeth(0, previous[i], 0));
```

Comment:
I recall reading that LLVM will never unroll loops which use the iteration variable in the body. A crate with macros to work around that has been created: https://docs.rs/crunchy/0.2.2/crunchy/

I wonder if this is relevant here?